apollo-runtime/src/main/java/com/apollographql/apollo/internal/RealApolloQueryWatcher.java

```diff
@@ -24,10 +25,10 @@
 
 final class RealApolloQueryWatcher<T> implements ApolloQueryWatcher<T> {
   private RealApolloCall<T> activeCall;
-  private ApolloCall.Callback<T> callback;
+  private Optional<ApolloCall.Callback<T>> callback = Optional.absent();
   private ResponseFetcher refetchResponseFetcher = ApolloResponseFetchers.CACHE_FIRST;
-  private volatile boolean canceled;
-  private boolean executed = false;
+  private boolean canceled;
+  private boolean executed;
```

Comment:
I would do the same as we did in `RealApolloCall` with atomic boolean and CAS operation